File: src/python/core/src/dicebear/prng/fnv1a.py
"""FNV-1a 32-bit hash."""

from __future__ import annotations

import struct


class Fnv1a:
    """FNV-1a 32-bit hash.

    Offset basis: ``0x811c9dc5``, prime: ``0x01000193``.

    See https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function
    """

    @staticmethod
    def hash(value: str) -> int:
        """Return the unsigned 32-bit FNV-1a hash of ``value``.

        The input is decomposed into UTF-16 code units before hashing so the
        result is identical across language ports.
        """
        result = 0x811C9DC5

        for code in _utf16_code_units(value):
            result = ((result ^ code) * 0x01000193) & 0xFFFFFFFF

        return result

    @staticmethod
    def hex(value: str) -> str:
        """Return the FNV-1a hash of ``value`` as an 8-char lowercase hex string."""
        return format(Fnv1a.hash(value), "08x")


def _utf16_code_units(value: str) -> list[int]:
    """Convert a string to its UTF-16 code units, matching JS ``charCodeAt``.

    Iterating ``value`` directly would yield code points and break on non-BMP
    input (e.g. emoji), so encode to little-endian UTF-16 and read uint16 units.
    """
    if value == "":
        return []

    encoded = value.encode("utf-16-le")

    return list(struct.unpack(f"<{len(encoded) // 2}H", encoded))

File: src/csharp/core/src/Prng/Fnv1a.cs
using System.Globalization;

namespace DiceBear.Internal
{
    /// <summary>
    /// FNV-1a 32-bit hash. Offset basis: 0x811c9dc5, prime: 0x01000193.
    /// </summary>
    /// <remarks>
    /// See https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function
    /// </remarks>
    internal static class Fnv1a
    {
        /// <summary>
        /// Returns the unsigned 32-bit FNV-1a hash of <paramref name="input"/>.
        /// </summary>
        /// <remarks>
        /// A .NET string is a sequence of UTF-16 code units, so iterating its
        /// chars is exactly what <c>charCodeAt</c> yields in the reference, and
        /// the hash is identical across the ports even for non-ASCII or non-BMP
        /// seeds. Unchecked <c>uint</c> arithmetic wraps on overflow,
        /// reproducing <c>Math.imul</c> and <c>&gt;&gt;&gt; 0</c>.
        /// </remarks>
        internal static uint Hash(string input)
        {
            var hash = 0x811c9dc5u;

            unchecked
            {
                foreach (var code in input)
                {
                    hash ^= code;
                    hash *= 0x01000193u;
                }
            }

            return hash;
        }

        /// <summary>
        /// Returns the FNV-1a hash of <paramref name="input"/> as an
        /// 8-character lowercase hex string.
        /// </summary>
        internal static string Hex(string input) =>
            Hash(input).ToString("x8", CultureInfo.InvariantCulture);
    }
}
